### PR TITLE
[CARBONDATA-1537] Fixed version compatabilty issues from V1 to latest carbon version

### PR DESCRIPTION
**Pull request by ravipesala**

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/218/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/342/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/970/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/343/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/219/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/220/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/971/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/344/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/972/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/345/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/221/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/346/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/222/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/973/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/974/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/223/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/347/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/975/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/224/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/348/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/976/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/349/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/225/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/977/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/350/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/226/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/978/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/228/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/352/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/980/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/353/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/229/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/981/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/354/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/230/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/982/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/355/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/231/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/983/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/232/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/356/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/984/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/357/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/233/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/985/

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java [hunk 1]
```diff
@@ -214,6 +218,42 @@ public void setNumberOfPages(int numberOfPages) {
     for (int i = 0; i < mSize; i++) {
       output.writeInt(measureChunksLength.get(i));
     }
+    // Serialize datachunks as well for older versions like V1 and V2
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: Can you wrap this logic into a function and mentioning it is for V1 and V2 serialization only, I think it will be more readable

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java [hunk 2]
```diff
@@ -238,6 +278,20 @@ public void setNumberOfPages(int numberOfPages) {
     for (int i = 0; i < measureChunkOffsetsSize; i++) {
       measureChunksLength.add(input.readInt());
     }
-
+    // Deserialize datachunks as well for older versions like V1 and V2
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: Can you wrap this logic into a function and mentioning it is for V1 and V2 deserialization only, I think it will be more readable

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java [hunk 3]
```diff
@@ -34,11 +34,13 @@
   STRUCT(10, "STRUCT", -1),
   MAP(11, "MAP", -1),
   BYTE(12, "BYTE", 1),
-
   // internal use only, for variable length data type
   BYTE_ARRAY(13, "BYTE_ARRAY", -1),
+
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java]: remove empty line

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java [hunk 4]
```diff
@@ -34,11 +34,13 @@
...
   // internal use only, for value compression from integer/long to 3 bytes value
-  SHORT_INT(14, "SHORT_INT", 3);
+  SHORT_INT(14, "SHORT_INT", 3),
+  // Only for internal use for backward compatability
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java]: mentioning it is for V1 and V2 format only

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 5]
```diff
@@ -2070,6 +1994,20 @@ public static void dropDatabaseDirectory(String dbName, String storePath)
     }
   }
 
+  public static DataType getDataType(char type) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Suggest to move all conversion to DataType enum, including this one and `ColumnPageEncoderMeta.convertType`. And also move `CarbonCommonConstants.BIG_INT_MEASURE` and related constants to DataType enum also.

File: core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverter.java [hunk 6]
```diff
@@ -123,6 +123,25 @@ private BlockletInfo getBlockletInfo(
   }
 
   @Override public List<ColumnSchema> getSchema(TableBlockInfo tableBlockInfo) throws IOException {
-    return null;
+    FileHolder fileReader = null;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverter.java]: Is this change related to this PR? Can you add comment to describe the intention of this function in its abstract interface?

File: integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CarbonV1toV3CompatabilityTestCase.scala [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.spark.sql.hive.CarbonSessionState
+import org.apache.spark.sql.test.TestQueryExecutor
+import org.apache.spark.sql.{CarbonEnv, CarbonSession, Row, SparkSession}
+import org.scalatest.BeforeAndAfterAll
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+
+/**
+ * V1 to V3 compatability test. This test has to be at last
```
Comment (jackylk) [inline integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CarbonV1toV3CompatabilityTestCase.scala]: Will you add V2 to V3 compatibility test in future PR?

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonSession.scala [hunk 8]
```diff
@@ -43,7 +43,7 @@ class CarbonSession(@transient val sc: SparkContext,
   }
 
   @transient
-  override private[sql] lazy val sessionState: SessionState = new CarbonSessionState(this)
+  override lazy val sessionState: SessionState = new CarbonSessionState(this)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonSession.scala]: Is this change required? Somewhere outside sql package will read this variable?

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: ok

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java (hunk 2 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: ok

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java (hunk 3 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java]: ok

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java (hunk 4 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DataType.java]: ok

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 5 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: ok

File: core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverter.java (hunk 6 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverter.java]: yes, it is part of this PR. We need to get columnschema

File: integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CarbonV1toV3CompatabilityTestCase.scala (hunk 7 above)
Comment (ravipesala) [reply integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CarbonV1toV3CompatabilityTestCase.scala]: yes, i will tests for V2 to V3 in another PR, Because I need to verify the compatibility of V2 to V3 first

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonSession.scala (hunk 8 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonSession.scala]: Yes, it is required for test case I added.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/294/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/418/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1048/

**Comment by jackylk** [comment]
LGTM
